File: runtime/go/overt/process.go
package overt

import (
	"bytes"
	"os"
	"os/exec"
)

// ProcessOutput is the captured result of a synchronous Process.run
// invocation: exit code plus stdout / stderr as strings. Field names
// are TitleCase per Go's exported-name convention; the Go emitter's
// stdlib-record field-access translation maps Overt's lowercase
// field references (`output.exit_code`) to the matching capitalized
// Go fields.
type ProcessOutput struct {
	ExitCode int
	Stdout   string
	Stderr   string
}

// ProcessRun runs cmd with the given args, blocks until it completes,
// and returns the captured outputs. A process that fails to launch
// surfaces as Err(IoError); a process that ran and exited non-zero
// is still Ok — the caller branches on output.exit_code.
func ProcessRun(cmd string, args List[string]) Result[ProcessOutput, IoError] {
	c := exec.Command(cmd, args.Items...)
	var stdoutBuf, stderrBuf bytes.Buffer
	c.Stdout = &stdoutBuf
	c.Stderr = &stderrBuf
	err := c.Run()
	// A non-zero exit surfaces as *exec.ExitError; the process did
	// run, just unhappily. Other error shapes (binary not found,
	// permission denied launching) are launch failures and surface
	// as Err.
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			return Ok[ProcessOutput, IoError](ProcessOutput{
				ExitCode: exitErr.ExitCode(),
				Stdout:   stdoutBuf.String(),
				Stderr:   stderrBuf.String(),
			})
		}
		return Err[ProcessOutput, IoError](IoError{Narrative: err.Error()})
	}
	return Ok[ProcessOutput, IoError](ProcessOutput{
		ExitCode: 0,
		Stdout:   stdoutBuf.String(),
		Stderr:   stderrBuf.String(),
	})
}

// Args returns the process command-line arguments minus the executable
// path that os.Args puts at index 0. Mirrors the C# runtime's Prelude.args()
// — both targets observe the same shape so a program reading argv via
// `args()` stdlib gets identical behavior across back ends. Returns the
// empty List when there are no user-supplied args. Effect-row-tracked
// `!{io}` because it observes process state.
func Args() List[string] {
	raw := os.Args
	if len(raw) <= 1 {
		return List[string]{Items: []string{}}
	}
	out := make([]string, len(raw)-1)
	copy(out, raw[1:])
	return List[string]{Items: out}
}

File: src/Overt.Runtime/Process.cs
// Process — synchronous subprocess execution. The v1 surface is one
// `run` operation that captures stdout / stderr / exit code.

namespace Overt.Runtime;

/// <summary>
/// The captured result of a synchronous Process.run invocation:
/// exit code plus stdout and stderr as strings. Field names match
/// Overt's lowercase-field convention so destructuring on the Overt
/// side reads naturally (`output.exit_code`, etc.).
/// </summary>
public sealed record ProcessOutput(int exit_code, string stdout, string stderr);

/// <summary>
/// Process companion. The v1 surface is one synchronous `run` operation
/// that captures stdout / stderr / exit code in full. Streaming I/O,
/// process groups, signals, and timeouts are deferred until a real
/// orchestration program needs them. Pairs with File / Path for the
/// minimum stdlib surface a CLI tool / build script / orchestrator
/// needs.
/// </summary>
public static class Process
{
    /// <summary>Run <paramref name="cmd"/> with the given <paramref name="args"/>,
    /// wait for it to complete, and return the captured stdout, stderr,
    /// and exit code. Failures to launch the process surface as
    /// <c>Err(IoError)</c>; a process that ran but exited non-zero is
    /// still <c>Ok</c> — callers branch on <c>output.exit_code</c>.</summary>
    public static Result<ProcessOutput, IoError> run(string cmd, List<string> args)
    {
        try
        {
            var psi = new global::System.Diagnostics.ProcessStartInfo(cmd)
            {
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
            };
            foreach (var a in args.Items) psi.ArgumentList.Add(a);
            using var p = global::System.Diagnostics.Process.Start(psi);
            if (p is null)
            {
                return new ResultErr<ProcessOutput, IoError>(
                    new IoError($"failed to start process: {cmd}"));
            }
            var stdout = p.StandardOutput.ReadToEnd();
            var stderr = p.StandardError.ReadToEnd();
            p.WaitForExit();
            return new ResultOk<ProcessOutput, IoError>(
                new ProcessOutput(p.ExitCode, stdout, stderr));
        }
        catch (Exception ex)
        {
            return new ResultErr<ProcessOutput, IoError>(new IoError(ex.Message));
        }
    }
}
